### PR TITLE
Arbitrary logic

### PR DESCRIPTION
**Pull request by jtremback**
This lands the `submitLogicCall` endpoint, along with a a logic batch contract and some benchmarking.

`submitLogicCall` allows you to make one arbitrary call to another contract from Peggy. It also transfers any number of fees in any number of tokens to the logic contract, and transfers any number of fees in any number of tokens to msg.sender. The fees differ from `submitBatch` in that they are expected to be computed beforehand, unlike `submitBatch`, which adds up the fees for each tx. This saves gas and increases flexibility.

`submitLogicCall` will usually be used with a logic batching contract, which executes a number of calls out to the actual logic contracts. This allows flexible upgrading of the batching logic. One of these batching contracts is included and tested in this PR.

`submitLogicCall` can also act as an extension point to regular token batches. There's another test which benchmarks a `submitLogicCall` token batch against a normal `submitBatch` token batch. `submitLogicCall` is slightly cheaper.

### CODE REVIEW

File: solidity/contracts/Peggy.sol [hunk 1]
```diff
@@ -343,6 +347,152 @@ contract Peggy {
...
+		// These are arrays of the parts of the validators signatures
+		uint8[] memory _v,
+		bytes32[] memory _r,
+		bytes32[] memory _s,
+		LogicCallArgs memory _args
+	) public {
+		// CHECKS scoped to reduce stack depth
+		{
+			// Check that the call has not timed out
+			require(block.timestamp < _args.timeOut, "Timed out");
```
Comment (jkilpatr) [inline solidity/contracts/Peggy.sol:390]: can we be consistent and use block numbers as established in #170 there are no real fundamental advantages to using timestamps.

**Review by jkilpatr — CHANGES_REQUESTED** [review]
Looks pretty good. It does look like the package.json for the solidity build tooling needs to be updated though.

File: solidity/contracts/Peggy.sol (hunk 1 above)
Comment (jtremback) [reply solidity/contracts/Peggy.sol:390]: Oh yea sorry i knew i was forgetting something

**Comment by jtremback** [comment]
@jkilpatr take a look, seems like github wants me to merge it though before running CI. I did get the all up tests working on an aws machine though

**Review by jkilpatr — APPROVED** [review]

**Pull request merged**